### DISCUSSION TITLE
Comparison to nfs-subdir-external-provisioner?

### DISCUSSION THREAD

**Discussion by c-x-berger** [kubernetes-csi/csi-driver-nfs#504] [Q&A]
Are there any practical differences between this project and the [`nfs-subdir-external-provisioner`](https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner) (which is also an apparently-official-sort-of k8s project)?

As far as I can tell, both of them are setting up "volumes" to fill `PersistentVolumClaims` as directories on an external NFS server. I'm not sure then (other than implementation details that I don't _think_ will be visible in my application) what differences there are between the two. This repo has "CSI" written all over it, but I'm not totally sure what that is or indeed why I should (not) care.

I realize this is a pretty basic question, but I'd appreciate if someone could help me understand how these two projects differ so I can pick the best one for my application.

**Comment by vaskozl** [comment]
I think this is a good question and wondered the same thing when I first saw the project. 

[nfs-subdir-external-provisioner](https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner) use the in-tree NFS support, meaning it just creates PV's with `.spec.nfs` populated. Kubernetes natively supported many different kinds of stroage (Ceph RBD, AWS EBS and other cloud specific storage). That supports is now getting moved out of core kubernetes into "drivers" which projects can support on their own, where the CSI standard says what the control plane will require of you. In simple terms the CSI driver calls `mount` instead of core kubernetes/kubelet itself.

You can read more about the [in-tree to CSI migration on the kubernetes.io blog](https://kubernetes.io/blog/2022/09/26/storage-in-tree-to-csi-migration-status-update-1.25/).

Now as it happens the in-tree support for NFS is not being removed so what advantage might this project provide?

* [Snapshots support with the snapshot-controller](https://kubernetes.io/docs/concepts/storage/volume-snapshots/) - the CSI design allows for `VolumeSnapshot` support, enabling backup solutions like kasten, volsync or restic to work  (N.B. snaphots are currently implemented via `tar` rather than FS/cow snapshots)
* Future extra features, e.g. [authentication](https://github.com/kubernetes-csi/csi-driver-nfs/issues/525), it's unlikely that the in-tree NFS support will allow anything more complicated

If you don't care about VolumeSnapshot's or new features which might come along you can stick to [nfs-subdir-external-provisioner](https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner)
